xxhash_unsafe.go

```diff
@@ -10,37 +10,60 @@ import (
 	"unsafe"
 )
 
-// Notes:
-//
-// See https://groups.google.com/d/msg/golang-nuts/dcjzJy-bSpw/tcZYBzQqAQAJ
-// for some discussion about these unsafe conversions.
-//
 // In the future it's possible that compiler optimizations will make these
-// unsafe operations unnecessary: https://golang.org/issue/2205.
-//
-// Both of these wrapper functions still incur function call overhead since they
-// will not be inlined. We could write Go/asm copies of Sum64 and Digest.Write
-// for strings to squeeze out a bit more speed. Mid-stack inlining should
-// eventually fix this.
+// XxxString functions unnecessary by realizing that calls such as
+// Sum64([]byte(s)) don't need to copy s. See https://golang.org/issue/2205.
+// If that happens, even if we keep these functions they can be replaced with
+// the trivial safe code.
 
 // Sum64String computes the 64-bit xxHash digest of s.
 // It may be faster than Sum64([]byte(s)) by avoiding a copy.
 func Sum64String(s string) uint64 {
 	var b []byte
-	bh := (*reflect.SliceHeader)(unsafe.Pointer(&b))
-	bh.Data = (*reflect.StringHeader)(unsafe.Pointer(&s)).Data
-	bh.Len = len(s)
-	bh.Cap = len(s)
+
+	// NOTE: The usual way of doing the conversion is as follows:
+	//
+	// bh := (*reflect.SliceHeader)(unsafe.Pointer(&b))
+	// bh.Data = (*reflect.StringHeader)(unsafe.Pointer(&s)).Data
+	// bh.Len = len(s)
+	// bh.Cap = len(s)
+	//
+	// Unfortunately, as of Go 1.15.3 the inliner considers this slightly
+	// too expensive to be an inlining candidate. Writing the conversion
+	// this way gets us down to the max cost (barely).
+
+	*(*string)(unsafe.Pointer(&b)) = s
+	(*sliceHeader)(unsafe.Pointer(&b)).cap = len(s)
```

Comment:
I suppose this should work as long as the memory layout of []byte is exactly that of sliceHeader (unsafe.Pointer docs promise (1)). The only way I can see this break is if the compiler or GC would at some point start checking that cap(b) >= len(b), which would also break the existing code (but that would be easier to fix).